File: mobile-app/src/services/dish.ts
import api from './api';
import { OSS_PUBLIC_BASE_URL } from '../config';

export interface CategoryInfo {
  id: number;
  name: string;
  sort: number;
}

export interface DishInfo {
  id: number;
  category_id: number;
  name: string;
  description: string;
  image_key: string;
  sort: number;
  created_at: string;
  updated_at: string;
}

export interface IngredientInfo {
  id: number;
  name: string;
  amount: number | null;
  amount_text: string;
  unit: string;
  sort: number;
}

export interface IngredientGroup {
  name: string;
  ingredients: IngredientInfo[];
}

export interface DishStepInfo {
  id: number;
  step_no: number;
  description: string;
  image_key: string;
  image_keys?: string[];
  timer_seconds?: number;
}

export interface DishDetailInfo {
  id: number;
  category_id: number;
  name: string;
  description: string;
  image_key: string;
  cook_minutes: number;
  difficulty: number;
  servings: number;
  tips: string;
  ingredient_groups: IngredientGroup[];
  steps: DishStepInfo[];
}

interface ListCategoriesResponse {
  code: number;
  message: string;
  categories: CategoryInfo[];
}

interface ListDishesResponse {
  code: number;
  message: string;
  dishes: DishInfo[];
}

interface DishDetailResponse {
  code: number;
  message: string;
  dish: DishDetailInfo;
}

/**
 * 查询所有分类
 */
export async function fetchCategories(): Promise<CategoryInfo[]> {
  const { data } = await api.get<ListCategoriesResponse>('/dish/categories');
  if (data.code !== 0) {
    throw new Error(data.message || '查询分类失败');
  }
  return data.categories ?? [];
}

/**
 * 查询某分类下的菜谱
 */
export async function fetchDishesByCategory(
  categoryId: number,
): Promise<DishInfo[]> {
  const { data } = await api.get<ListDishesResponse>('/dish/dishes', {
    params: { category_id: categoryId },
  });
  if (data.code !== 0) {
    throw new Error(data.message || '查询菜谱失败');
  }
  return data.dishes ?? [];
}

/**
 * 搜索菜谱
 */
export async function searchDishes(keyword: string): Promise<DishInfo[]> {
  const { data } = await api.get<ListDishesResponse>('/dish/search', {
    params: { keyword },
  });
  if (data.code !== 0) {
    throw new Error(data.message || '搜索失败');
  }
  return data.dishes ?? [];
}

/**
 * 查询菜谱详情
 */
export async function fetchDishDetail(dishId: number): Promise<DishDetailInfo> {
  const { data } = await api.get<DishDetailResponse>(`/dish/${dishId}`);
  if (data.code !== 0 || !data.dish) {
    throw new Error(data.message || '查询菜谱详情失败');
  }
  return {
    ...data.dish,
    ingredient_groups: data.dish.ingredient_groups ?? [],
    steps: data.dish.steps ?? [],
  };
}

/**
 * 将 OSS key 转成可访问的公网 URL；完整 URL 原样返回。
 */
export function resolveDishImageUrl(imageKey: string): string | null {
  const key = imageKey.trim();
  if (!key) return null;
  if (/^https?:\/\//i.test(key)) return key;
  return `${OSS_PUBLIC_BASE_URL}/${key.replace(/^\/+/, '')}`;
}

/**
 * 兼容当前单图 image_key 和后续多图 image_keys。
 */
export function resolveStepImageUrls(step: DishStepInfo): string[] {
  const keys = step.image_keys?.length
    ? step.image_keys
    : step.image_key
      ? [step.image_key]
      : [];

  return Array.from(
    new Set(
      keys
        .map(resolveDishImageUrl)
        .filter((url): url is string => Boolean(url)),
    ),
  );
}

File: gateway-service/internal/handler/dish.go
package handler

import (
	"context"
	"log"
	"net/http"
	"strconv"
	"time"

	"github.com/gin-gonic/gin"
	dishv1 "github.com/lijunsheng/familyos/proto/gen/dish/v1"
	"google.golang.org/grpc"
)

// DishHandler 菜谱 HTTP 处理器
type DishHandler struct {
	client dishv1.DishServiceClient
}

// NewDishHandler 创建 DishHandler
func NewDishHandler(conn *grpc.ClientConn) *DishHandler {
	return &DishHandler{
		client: dishv1.NewDishServiceClient(conn),
	}
}

// ListCategories GET /api/v1/dish/categories
func (h *DishHandler) ListCategories(c *gin.Context) {
	ctx, cancel := context.WithTimeout(c.Request.Context(), 5*time.Second)
	defer cancel()

	resp, err := h.client.ListCategories(ctx, &dishv1.ListCategoriesRequest{})
	if err != nil {
		log.Printf("gRPC ListCategories 调用失败: %v", err)
		c.JSON(http.StatusInternalServerError, gin.H{
			"code":    1999,
			"message": "服务内部错误",
		})
		return
	}

	categories := make([]gin.H, 0, len(resp.GetCategories()))
	for _, cat := range resp.GetCategories() {
		categories = append(categories, gin.H{
			"id":   cat.GetId(),
			"name": cat.GetName(),
			"sort": cat.GetSort(),
		})
	}

	c.JSON(http.StatusOK, gin.H{
		"code":       resp.GetCode(),
		"message":    resp.GetMessage(),
		"categories": categories,
	})
}

// ListDishesByCategory GET /api/v1/dish/dishes?category_id=1
func (h *DishHandler) ListDishesByCategory(c *gin.Context) {
	categoryIDStr := c.DefaultQuery("category_id", "0")
	categoryID, err := strconv.ParseInt(categoryIDStr, 10, 64)
	if err != nil || categoryID <= 0 {
		c.JSON(http.StatusBadRequest, gin.H{
			"code":    1999,
			"message": "请提供有效的 category_id",
		})
		return
	}

	ctx, cancel := context.WithTimeout(c.Request.Context(), 5*time.Second)
	defer cancel()

	req := &dishv1.ListDishesByCategoryRequest{
		CategoryId: categoryID,
	}
	resp, err := h.client.ListDishesByCategory(ctx, req)
	if err != nil {
		log.Printf("gRPC ListDishesByCategory 调用失败: %v", err)
		c.JSON(http.StatusInternalServerError, gin.H{
			"code":    1999,
			"message": "服务内部错误",
		})
		return
	}

	dishes := make([]gin.H, 0, len(resp.GetDishes()))
	for _, d := range resp.GetDishes() {
		dishes = append(dishes, gin.H{
			"id":          d.GetId(),
			"category_id": d.GetCategoryId(),
			"name":        d.GetName(),
			"description": d.GetDescription(),
			"image_key":   d.GetImageKey(),
			"sort":        d.GetSort(),
			"created_at":  d.GetCreatedAt(),
			"updated_at":  d.GetUpdatedAt(),
		})
	}

	c.JSON(http.StatusOK, gin.H{
		"code":    resp.GetCode(),
		"message": resp.GetMessage(),
		"dishes":  dishes,
	})
}

// SearchDishes GET /api/v1/dish/search?keyword=红烧
func (h *DishHandler) SearchDishes(c *gin.Context) {
	keyword := c.DefaultQuery("keyword", "")
	if keyword == "" {
		c.JSON(http.StatusBadRequest, gin.H{
			"code":    1999,
			"message": "请提供搜索关键字",
		})
		return
	}

	ctx, cancel := context.WithTimeout(c.Request.Context(), 5*time.Second)
	defer cancel()

	req := &dishv1.SearchDishesRequest{Keyword: keyword}
	resp, err := h.client.SearchDishes(ctx, req)
	if err != nil {
		log.Printf("gRPC SearchDishes 调用失败: %v", err)
		c.JSON(http.StatusInternalServerError, gin.H{
			"code":    1999,
			"message": "服务内部错误",
		})
		return
	}

	dishes := make([]gin.H, 0, len(resp.GetDishes()))
	for _, d := range resp.GetDishes() {
		dishes = append(dishes, gin.H{
			"id":          d.GetId(),
			"category_id": d.GetCategoryId(),
			"name":        d.GetName(),
			"description": d.GetDescription(),
			"image_key":   d.GetImageKey(),
			"sort":        d.GetSort(),
			"created_at":  d.GetCreatedAt(),
			"updated_at":  d.GetUpdatedAt(),
		})
	}

	c.JSON(http.StatusOK, gin.H{
		"code":    resp.GetCode(),
		"message": resp.GetMessage(),
		"dishes":  dishes,
	})
}

// GetDishDetail GET /api/v1/dish/:id
func (h *DishHandler) GetDishDetail(c *gin.Context) {
	dishID, err := strconv.ParseInt(c.Param("id"), 10, 64)
	if err != nil || dishID <= 0 {
		c.JSON(http.StatusBadRequest, gin.H{
			"code":    2001,
			"message": "菜谱ID不正确",
		})
		return
	}

	ctx, cancel := context.WithTimeout(c.Request.Context(), 5*time.Second)
	defer cancel()

	resp, err := h.client.GetDishDetail(ctx, &dishv1.GetDishDetailRequest{DishId: dishID})
	if err != nil {
		log.Printf("gRPC GetDishDetail 调用失败: %v", err)
		c.JSON(http.StatusInternalServerError, gin.H{
			"code":    1999,
			"message": "服务内部错误",
		})
		return
	}
	if resp.GetCode() == 2002 {
		c.JSON(http.StatusNotFound, gin.H{
			"code":    resp.GetCode(),
			"message": resp.GetMessage(),
		})
		return
	}
	if resp.GetCode() != 0 || resp.GetDish() == nil {
		c.JSON(http.StatusInternalServerError, gin.H{
			"code":    resp.GetCode(),
			"message": resp.GetMessage(),
		})
		return
	}

	dish := resp.GetDish()
	groups := make([]gin.H, 0, len(dish.GetIngredientGroups()))
	for _, group := range dish.GetIngredientGroups() {
		ingredients := make([]gin.H, 0, len(group.GetIngredients()))
		for _, ingredient := range group.GetIngredients() {
			var amount any
			if ingredient.Amount != nil {
				amount = ingredient.GetAmount()
			}
			ingredients = append(ingredients, gin.H{
				"id":          ingredient.GetId(),
				"name":        ingredient.GetName(),
				"amount":      amount,
				"amount_text": ingredient.GetAmountText(),
				"unit":        ingredient.GetUnit(),
				"sort":        ingredient.GetSort(),
			})
		}
		groups = append(groups, gin.H{
			"name":        group.GetName(),
			"ingredients": ingredients,
		})
	}

	steps := make([]gin.H, 0, len(dish.GetSteps()))
	for _, step := range dish.GetSteps() {
		steps = append(steps, gin.H{
			"id":          step.GetId(),
			"step_no":     step.GetStepNo(),
			"description": step.GetDescription(),
			"image_key":   step.GetImageKey(),
		})
	}

	c.JSON(http.StatusOK, gin.H{
		"code":    resp.GetCode(),
		"message": resp.GetMessage(),
		"dish": gin.H{
			"id":                dish.GetId(),
			"category_id":       dish.GetCategoryId(),
			"name":              dish.GetName(),
			"description":       dish.GetDescription(),
			"image_key":         dish.GetImageKey(),
			"cook_minutes":      dish.GetCookMinutes(),
			"difficulty":        dish.GetDifficulty(),
			"servings":          dish.GetServings(),
			"tips":              dish.GetTips(),
			"ingredient_groups": groups,
			"steps":             steps,
		},
	})
}
